### ISSUE TITLE
bunnies are sometimes partially outside the environment bounds

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/natural-selection#131]
Noticed in [1.0.0-dev.70](https://phet-dev.colorado.edu/html/natural-selection/1.0.0-dev.70/phet/natural-selection_en_phet.html).

Very occasionally, a bunny will get a tiny bit outside of the environment bounds. For example:

<img width="969" alt="screenshot_432" src="https://user-images.githubusercontent.com/3046552/87630848-598d3380-c6f3-11ea-9762-d5d6a8006633.png">
 
This occurs when a bunny hops towards the edge of the bounds and changes direction.  This might involve tweaking `X_MARGIN` in Bunny.js, which determines how close a bunny can get to the edges of the bounds.    Or this might be a bug in Bunny.js `initializeMotion`.

Note that switching to scenery Sprites in #128 means that we can no longer use `clipArea`.  So we can't rely on clipping for this.

**Comment by pixelzoom** [comment]
Increasing X_MARGIN from 25 to 28 seems to do the job.  I tested with `?initialScreen=2&labPopulation=249&speed=2`, watched the bunnies hop around for a long time, and didn't see any leave the environment area.

Closing.

**Comment by phet-steele** [comment]
I believe this is still happening in 1.3.0-dev.1.

![image](https://user-images.githubusercontent.com/5863899/114247125-9a983880-9951-11eb-9af4-48385ad1f2f1.png)

I tested with `?initialScreen=2&labPopulation=2000&speed=8` to get the above. Some attempts it happened quite quickly, some attempts it didn't. Not really sure the best way to reproduce other than just waiting.

**Comment by pixelzoom** [comment]
Good one. In addition to the bunny way over to the left, I see the bottom part of a bunny just below the "Add Mutations" panel.

**Comment by pixelzoom** [comment]
I believe that you're seeing this, but I haven't been able to reproduce it. I'm trying to evaluate the severity, whether it needs to be addressed for #251 (1.3 release).

@phet-steele when you see this occur, does the bunny that is outside the "environment" panel move back into that panel, or does it continue to wander around outside that panel?

**Comment by phet-steele** [comment]
> I believe that you're seeing this, but I haven't been able to reproduce it. I'm trying to evaluate the severity, whether it needs to be addressed for #251 (1.3 release).

Not sure if this will help, but a hunch says aspect ratio could influence it? Try to go wider and give the bunnies plenty of empty space on either side of the environment. I don't think this actually matters, but it is worth a shot if you are seeing nothing! I can do some more rigorous time trials to see how long it takes for me to see this, if you would like @pixelzoom.

> @phet-steele when you see this occur, does the bunny that is outside the "environment" panel move back into that panel, or does it continue to wander around outside that panel?

It wanders around aimlessly, and does not appear to tend towards the environment. They do sometimes head back into the environment, but that is about as random as them leaving it in the first place.

**Comment by pixelzoom** [comment]
By running with `?speed=8`, you're running the sim with a time step (dt) that is larger than a bunny's typical hop cycle.   In Bunny.js, this constant defined the range for the time of the hop cycle, and a value is chosen randomly from this range for each hop:

```js
const HOP_TIME_RANGE = new Range( 0.25, 0.5 ); // time to complete a hop cycle, in seconds
```

On my MacBook Pro for example, `?speed=8` results in `dt` values of ~0.4 seconds, which is greater than some hop cycles. This will cause bunnies to "over hop", possibly leaving the bounds of the panel.

@phet-steele have you every observed this problem with `speed=1`?

**Comment by phet-steele** [comment]
> @phet-steele have you every observed this problem with `speed=1`? (...where a typical `dt` is ~0.05 sec.)

Yes, I first saw it in the Studio wrapper at normal speed, which is what prompted me to attempt to reproduce. I used your fancy query params to speed that up! 😁

**Comment by pixelzoom** [comment]
> Not sure if this will help, but a hunch says aspect ratio could influence it? Try to go wider and give the bunnies plenty of empty space on either side of the environment. I don't think this actually matters, but it is worth a shot if you are seeing nothing! 

Thanks for the suggestion. But there's nothing in sim-specific code that cares about the dimensions of the browser window.   All of the bunny and wolf motion is based on a trapezoid that defines the 3D space inside the environment panel.  And that doesn't change dimensions when the browser window is resized.

But resizing the browser does make it more likely to see bunnies that have escaped.

> It wanders around aimlessly, and does not appear to tend towards the environment. They do sometimes head back into the environment, but that is about as random as them leaving it in the first place.

That's what I was afraid of. At the very least, it should be possible to give them a giant knudge into the environment panel.

**Comment by pixelzoom** [comment]
I said:
>  I'm trying to evaluate the severity, whether it needs to be addressed for #251 (1.3 release).

@kathy-phet @amanda-phet is this issue blocking for #251 (1.3 release)?  I've spent 4 hours on it so far, and it's proving to be very difficult.

**Comment by kathy-phet** [comment]
4 hours is a lot of time. It seems like a rare occurrence and since it doesn't break anything or cause a crash, I would vote for "wontfix".

**Comment by pixelzoom** [comment]
@kathy-phet said:
> ... I would vote for "wontfix".

Sorry, I didn't get your reply until late in the game, and I've fixed it.

Summary of the changes:
* Major changes to the `initializeMotion` and `hop` methods in Bunny.js.
* Detect when a bunny is unreasonably far from the trapezoid that defines the environment bounds. A little bit outside the trapezoid along the x axis is OK, and the bunny will still appear to be inside the environment panel.  Too far outside will trigger an assertion failure.
* If a bunny does get outside the environment panel when running without assertions, it will find its way back. It will not "wander around aimlessly", but will instead return to the panel.
* Modified BunnyIO to handle the above revisions, which required new state information.
* Prevent large dt values (like with `?speed=8`) from misbehaving.

I fuzz tested for ~20 minutes with the configuration shown in https://github.com/phetsims/natural-selection/issues/131#issuecomment-817011205, and I didn't see any bunnies that outside the environment panel.  I also tested in Studio and State Wrapper.

@phet-steele please verify with standalone, Studio, and State Wrapper. High priority please, since this is the last issue before creating the release branch for #251.

**Comment by kathy-phet** [comment]
Well glad its fixed!

**Comment by phet-steele** [comment]
This has seemed to be working well on master, so we will verify in rc!

**Comment by pixelzoom** [comment]
To verify this issue:

* Run the sim with `?initialScreen=2&labPopulation=100` and verify that the bunnies look like they are hopping around as expected. 
* Run the sim with `?initialScreen=2&labPopulation=2000&speed=8` and verify that the bunnies stay in the environment panel.

**Comment by phet-steele** [comment]
This has looked fixed in 1.3.0-rc.1! 👍
